### PR TITLE
Add ONNX Training Post-Passes to Front-End

### PR DESCRIPTION
**Pull request by lara-hdr**
**Description**: This PR adds ONNX post-process to the PyTorch front-end. The User can disable these passes by setting _enable_internal_postprocess to False (default is True).  

**Motivation and Context**
- These are common training post-passes, by ading them to the front-end, we abstract complexity from the user.

### CODE REVIEW

File: orttraining/orttraining/test/python/onnxruntime_test_postprocess.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+
+        onnx_model = self.get_onnx_model(model, model_desc, input_args, device)
+
+        count_layer_norm = self.count_nodes(onnx_model, "LayerNormalization")
+        count_nodes = self.count_all_nodes(onnx_model)
+       
+        assert count_layer_norm == 1
+        assert count_nodes == 3
+
+    def test_expand(self):
```
Comment (lara-hdr) [inline orttraining/orttraining/test/python/onnxruntime_test_postprocess.py:105]: In this test I the learning rate as the last input in input_args with the descriptor " IODescription('Learning_Rate', [1,], torch.float32)". But I still get a binding 'Learning_Rate' error. I follow the same logic in test_layer_norm an don't get the error. Any idea why this is happening?

**Comment by lara-hdr** [comment]
@thiagocrepaldi @BowenBao, let me know if there are any comments on the changes to _enable_internal_postprocess and _extra_postprocess flags.

**Comment by thiagocrepaldi** [comment]
> 
> 
> @thiagocrepaldi @BowenBao, let me know if there are any comments on the changes to _enable_internal_postprocess and _extra_postprocess flags.

Apparently your fork of ORT is outdated as it still have ORTModel, which was deleted from master. This PR will not merge due to conflicts. You will have to update your master branch of fork and rebase your private branch on it to merge it cleanly

File: orttraining/orttraining/python/ort_trainer.py [hunk 2]
```diff
@@ -681,6 +628,9 @@ def _init_session(self):
         if self.onnx_model_ is None:
             return
 
+        if self._enable_internal_postprocess:
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:631]: This postprocess cover the scenario where user provides onnx_model from the beginning

File: orttraining/orttraining/python/ort_trainer.py [hunk 3]
```diff
@@ -727,11 +677,13 @@ def _init_session(self):
 
     def _init_onnx_model(self, inputs):
         if self.onnx_model_ is not None:
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:679]: `_init_onnx_model` is called on `train_step` and `eval_step` only, so at this point, either 1) `self.onnx_model_` was already provided on `ORTTrainer` constructor and post processed by `self._init_session` (line 631) and it doesn't need to be postprocessed again (line 680-681 seems redundant) or 2) `self.torch_model_` was specified on `ORTTrainer` and will be processed by the `convert_model_loss_fn_to_onnx` call below (line 686)

File: orttraining/orttraining/python/ort_trainer.py [hunk 4]
```diff
@@ -655,7 +601,7 @@ def __init__(self, model, loss_fn, model_desc, training_optimizer_name, map_opti
         # we use self.global_step_ to count optimizations being performed.
         # it is used to calculate learning rate if self.get_lr_this_step_ is provided.
         self.global_step_ = global_step
-        self.post_process_model_fn_ = postprocess_model
+        self.post_process_model_fn_ = _extra_postprocess
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:604]: I suggest renaming `post_process_model_fn_ ` for consistency with the public flag `_extra_postprocess`. Makes debugging easy

File: orttraining/orttraining/python/ort_trainer.py [hunk 5]
```diff
@@ -681,6 +628,9 @@ def _init_session(self):
         if self.onnx_model_ is None:
             return
 
+        if self._enable_internal_postprocess:
+            self._onnx_model_ = postprocess.run_postprocess(self.onnx_model_)
+
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:633]: The call on lines 631-632 covers the scenario where user provides `self.onnx_model_` at `ORTTrainer` constructor. Note it only does internal post processing, but not extra post processing. This seems to be a pre-exsiting bug
Suggest adding the following on line 633
```python
if self.post_process_model_fn_:
                self.post_process_model_fn_(self.onnx_model_)
```

File: orttraining/orttraining/python/ort_trainer.py [hunk 6]
```diff
@@ -983,8 +935,8 @@ def _verify_fully_optimized_model(self, model):
             raise RuntimeError("the first output of a model to run with fully optimized ORT backend must be a scaler.")
 
 class ORTModel():
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:937]: This class was deleted on master. It will conflict during merge

**Comment by thiagocrepaldi** [comment]
Did you noticed that no CI Checks are being done on this PR? Is it because this is a fork and not a private branch on ORT? It would be nice to see CI passing after all these changes

File: orttraining/orttraining/python/ort_trainer.py [hunk 7]
```diff
@@ -320,15 +264,16 @@ def convert_model_loss_fn_to_onnx(model, loss_fn, model_desc, device, inputs, op
                        output_names=output_names,
                        opset_version=opset_version,
                        dynamic_axes=dynamic_axes,
-                       training=True,
+                       training=torch.onnx.TrainingMode.TRAINING,
```
Comment (BowenBao) [inline orttraining/orttraining/python/ort_trainer.py:267]: this does not work with older pt version

File: orttraining/orttraining/test/python/onnxruntime_test_postprocess.py [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+
+        device = torch.device("cpu")
+        target = torch.ones(20, 10, 10, dtype=torch.int64).to(device)
+        model = LayerNormNet(target)
+        input = torch.randn(20, 5, 10, 10, dtype=torch.float32).to(device)
+
+        input_desc = IODescription('input', [], "float32")
+        output0_desc = IODescription('output0', [], "float32")
+        output1_desc = IODescription('output1', [20, 5, 10, 10], "float32")
+        output2_desc = IODescription('output2', [20, 5, 10, 10], "float32")
```
Comment (BowenBao) [inline orttraining/orttraining/test/python/onnxruntime_test_postprocess.py:91]: nit: unused `output2_desc`
